sdks/java/core/src/test/java/org/apache/beam/sdk/options/ProxyInvocationHandlerTest.java

```diff
@@ -683,10 +722,233 @@ public void testJsonConversionOfSerializableWithMetadataProperty() throws Except
+  }
+
+  interface TypedOptions extends PipelineOptions {
+    int getInteger();
+    void setInteger(int value);
+
+    Instant getTimestamp();
+    void setTimestamp(Instant value);
+
+    Class<?> getJavaClass();
+    void setJavaClass(Class<?> value);
+
+    Object getObject();
+    void setObject(Object value);
+  }
+
+  @Test
+  public void testDisplayDataInheritanceNamespace() {
+    ExtendsBaseOptions options = PipelineOptionsFactory.as(ExtendsBaseOptions.class);
+    options.setFoo("bar");
+
+    DisplayData displayData = DisplayData.from(options);
+
+    assertThat(displayData, hasDisplayItem(allOf(
+        hasKey("foo"),
+        hasValue("bar"),
+        hasNamespace(ExtendsBaseOptions.class)
+    )));
+  }
+
+  interface BaseOptions extends PipelineOptions {
+    String getFoo();
+    void setFoo(String value);
+  }
+
+  interface ExtendsBaseOptions extends BaseOptions {
+    @Override String getFoo();
+    @Override void setFoo(String value);
+  }
+
+  @Test
+  public void testDisplayDataExcludedFromOverriddenBaseClass() {
+    ExtendsBaseOptions options = PipelineOptionsFactory.as(ExtendsBaseOptions.class);
+    options.setFoo("bar");
+
+    DisplayData displayData = DisplayData.from(options);
+    assertThat(displayData, not(hasDisplayItem(hasNamespace(BaseOptions.class))));
+  }
+
+  @Test
+  public void testDisplayDataIncludedForDisjointInterfaceHierarchies() {
+    FooOptions fooOptions = PipelineOptionsFactory.as(FooOptions.class);
+    fooOptions.setFoo("foo");
+
+    BarOptions barOptions = fooOptions.as(BarOptions.class);
+    barOptions.setBar("bar");
+
+    DisplayData data = DisplayData.from(barOptions);
+    assertThat(data, hasDisplayItem(allOf(hasKey("foo"), hasNamespace(FooOptions.class))));
+    assertThat(data, hasDisplayItem(allOf(hasKey("bar"), hasNamespace(BarOptions.class))));
+  }
+
+  interface FooOptions extends PipelineOptions {
+    String getFoo();
+    void setFoo(String value);
+  }
+
+  interface BarOptions extends PipelineOptions {
+    String getBar();
+    void setBar(String value);
+  }
+
+  @Test
+  public void testDisplayDataExcludesDefaultValues() {
+    PipelineOptions options = PipelineOptionsFactory.as(HasDefaults.class);
+    DisplayData data = DisplayData.from(options);
+
+    assertThat(data, not(hasDisplayItem(hasKey("foo"))));
+  }
+
+  interface HasDefaults extends PipelineOptions {
+    @Default.String("bar")
+    String getFoo();
+    void setFoo(String value);
+  }
+
+  @Test
+  public void testDisplayDataExcludesValuesAccessedButNeverSet() {
+    HasDefaults options = PipelineOptionsFactory.as(HasDefaults.class);
+    assertEquals("bar", options.getFoo());
+
+    DisplayData data = DisplayData.from(options);
+    assertThat(data, not(hasDisplayItem(hasKey("foo"))));
+  }
+
+  @Test
+  public void testDisplayDataIncludesExplicitlySetDefaults() {
+    HasDefaults options = PipelineOptionsFactory.as(HasDefaults.class);
+    options.setFoo("bar");
```

Comment:
options.setFoo(options.getFoo()) would test something that was default being overridden